brokerapi/brokers/spanner/broker.go

```diff
@@ -0,0 +1,280 @@
+	"net/http"
+	"strconv"
+)
+
+type SpannerBroker struct {
+	Client         *http.Client
+	ProjectId      string
+	Logger         lager.Logger
+	AccountManager models.AccountManager
+
+	broker_base.BrokerBase
+}
+
+type InstanceInformation struct {
+	InstanceId string `json:"instance_id"`
+}
+
+// Creates a new Spanner Instance identified by the name provided in details.RawParameters.name and
+// an optional region (defaults to regional-us-central1) and optional display_name
+func (s *SpannerBroker) Provision(instanceId string, details models.ProvisionDetails, plan models.PlanDetails) (models.ServiceInstanceDetails, error) {
+	var err error
+	var params map[string]string
+
+	if len(details.RawParameters) == 0 {
+		params = map[string]string{}
+	} else if err = json.Unmarshal(details.RawParameters, &params); err != nil {
+		return models.ServiceInstanceDetails{}, fmt.Errorf("Error unmarshalling parameters: %s", err)
+	}
+
+	// Ensure there is a name for this instance
+	if _, ok := params["name"]; !ok {
+		params["name"] = name_generator.Basic.InstanceNameWithSeparator("-")
+	}
+
+	// get plan parameters
+	var planDetails map[string]string
+	if err = json.Unmarshal([]byte(plan.Features), &planDetails); err != nil {
+		return models.ServiceInstanceDetails{}, fmt.Errorf("Error unmarshalling plan features: %s", err)
+	}
+
+	// set up client
+
+	client, err := googlespanner.NewInstanceAdminClient(context.Background(), option.WithUserAgent(models.CustomUserAgent))
+	if err != nil {
+		return models.ServiceInstanceDetails{}, fmt.Errorf("Error creating client: %s", err)
+	}
+
+	// set up params
+	numNodes, err := strconv.Atoi(planDetails["num_nodes"])
+	if err != nil {
+		return models.ServiceInstanceDetails{}, fmt.Errorf("Error getting number of nodes: %s", err)
+	}
+
+	displayName := params["name"]
+	if customerDisplayName, ok := params["display_name"]; ok {
+		displayName = customerDisplayName
+	}
+
+	loc, ok := params["location"]
+	if !ok {
+		loc = "projects/" + s.ProjectId + "/instanceConfigs/regional-us-central1"
+	} else {
+		loc = "projects/" + s.ProjectId + "/instanceConfigs/" + loc
+	}
+
+	// create instance
+	op, err := client.CreateInstance(context.Background(), &instancepb.CreateInstanceRequest{
+		Parent:     "projects/" + s.ProjectId,
+		InstanceId: params["name"],
+		Instance: &instancepb.Instance{
+			Name:        "projects/" + s.ProjectId + "/instances/" + params["name"],
+			DisplayName: displayName,
+			NodeCount:   int32(numNodes),
+			Config:      loc,
+		},
+	})
+	if err != nil {
+		return models.ServiceInstanceDetails{}, fmt.Errorf("Error creating instance: %s", err)
+	}
+
+	// save off instance information
+	ii := InstanceInformation{
+		InstanceId: params["name"],
+	}
+
+	otherDetails, err := json.Marshal(ii)
+	if err != nil {
+		return models.ServiceInstanceDetails{}, fmt.Errorf("Error marshalling other details: %s", err)
+	}
+
+	i := models.ServiceInstanceDetails{
```

Comment:
Do you think it'd be useful to move the update of `ServiceInstanceDetail` into `BrokerBase`? All of the brokers I am familiar with do this serialize otherDetails then persist this model.